File: system/loggerd/memory_pressure.py
#!/usr/bin/env python3
"""
Memory pressure handling utilities for openpilot.
Provides functions to check memory usage and clear caches when memory is low.
"""
import psutil
from openpilot.common.swaglog import cloudlog

# Memory thresholds
MEMORY_WARNING_THRESHOLD = 75  # Start taking action at 75%
MEMORY_CRITICAL_THRESHOLD = 80  # Critical threshold - same as controlsd
MEMORY_EMERGENCY_THRESHOLD = 85  # Emergency - clear everything possible

def get_memory_usage_percent() -> int:
  """Get current memory usage percentage."""
  return int(round(psutil.virtual_memory().percent))

def is_memory_pressure_low() -> bool:
  """Check if memory pressure is low (below warning threshold)."""
  return get_memory_usage_percent() < MEMORY_WARNING_THRESHOLD

def is_memory_pressure_high() -> bool:
  """Check if memory pressure is high (above warning threshold)."""
  return get_memory_usage_percent() >= MEMORY_WARNING_THRESHOLD

def is_memory_pressure_critical() -> bool:
  """Check if memory pressure is critical (above critical threshold)."""
  return get_memory_usage_percent() >= MEMORY_CRITICAL_THRESHOLD

def is_memory_pressure_emergency() -> bool:
  """Check if memory pressure is in emergency state (above emergency threshold)."""
  return get_memory_usage_percent() >= MEMORY_EMERGENCY_THRESHOLD

def clear_xattr_cache() -> int:
  """
  Clear the extended attributes cache.
  Returns number of entries cleared.
  """
  try:
    from openpilot.system.loggerd.xattr_cache import _cached_attributes
    cleared = len(_cached_attributes)
    _cached_attributes.clear()
    if cleared > 0:
      cloudlog.info(f"Cleared {cleared} xattr cache entries due to memory pressure")
    return cleared
  except Exception as e:
    cloudlog.error(f"Failed to clear xattr cache: {e}")
    return 0

def clear_filesystem_cache() -> None:
  """
  Attempt to clear filesystem page cache by dropping caches.
  Requires root privileges. This is a best-effort operation.
  """
  try:
    import subprocess
    # Write 1 to drop page cache (see /proc/sys/vm/drop_caches)
    # This requires root, so we try but don't fail if it doesn't work
    subprocess.run(
      ["sudo", "sh", "-c", "echo 1 > /proc/sys/vm/drop_caches"],
      check=False,
      capture_output=True,
      timeout=5
    )
    cloudlog.info("Attempted to clear filesystem page cache")
  except Exception as e:
    # Silently fail - this is optional
    pass

def handle_memory_pressure(clear_caches: bool = True, clear_fs_cache: bool = False) -> dict:
  """
  Handle memory pressure by clearing caches and returning status.
  
  Args:
    clear_caches: If True, clear application caches (xattr cache)
    clear_fs_cache: If True, attempt to clear filesystem page cache (requires root)
  
  Returns:
    dict with memory usage info and actions taken
  """
  mem_percent = get_memory_usage_percent()
  mem_info = psutil.virtual_memory()
  
  result = {
    "memory_percent": mem_percent,
    "memory_available_mb": mem_info.available // (1024 * 1024),
    "memory_total_mb": mem_info.total // (1024 * 1024),
    "xattr_cache_cleared": 0,
    "fs_cache_cleared": False,
  }
  
  if is_memory_pressure_high():
    if clear_caches:
      result["xattr_cache_cleared"] = clear_xattr_cache()
    
    if clear_fs_cache and is_memory_pressure_critical():
      clear_filesystem_cache()
      result["fs_cache_cleared"] = True
    
    cloudlog.warning(
      f"Memory pressure detected: {mem_percent}% used, "
      f"{result['memory_available_mb']}MB available. "
      f"Cleared {result['xattr_cache_cleared']} cache entries."
    )
  
  return result

def should_skip_filesystem_operation() -> bool:
  """
  Determine if filesystem operations should be skipped due to memory pressure.
  Returns True if memory is critical and operations should be skipped.
  """
  return is_memory_pressure_critical()

File: system/loggerd/memory_pressure.cc
#include "system/loggerd/memory_pressure.h"

#include <fstream>
#include <sstream>
#include <string>
#include <unordered_map>

#include "common/swaglog.h"
#include "common/timing.h"

namespace MemoryPressure {

static int parse_meminfo() {
  std::ifstream stream("/proc/meminfo");
  if (!stream.is_open()) {
    return -1;
  }

  std::unordered_map<std::string, uint64_t> mem_info;
  std::string line, key;
  
  while (std::getline(stream, line)) {
    uint64_t val = 0;
    std::istringstream iss(line);
    if (iss >> key >> val) {
      // Values in /proc/meminfo are in KB, convert to bytes
      mem_info[key] = val * 1024;
    }
  }

  if (mem_info.find("MemTotal:") == mem_info.end() || 
      mem_info.find("MemAvailable:") == mem_info.end()) {
    return -1;
  }

  uint64_t mem_total = mem_info["MemTotal:"];
  uint64_t mem_available = mem_info["MemAvailable:"];
  
  if (mem_total == 0) {
    return -1;
  }

  uint64_t mem_used = mem_total - mem_available;
  int percent = (mem_used * 100) / mem_total;
  
  return percent;
}

int get_memory_usage_percent() {
  static int cached_percent = -1;
  static double last_check_ms = 0;
  
  // Cache for 1 second to avoid reading /proc/meminfo too frequently
  double now_ms = millis_since_boot();
  if (cached_percent >= 0 && (now_ms - last_check_ms) < 1000) {
    return cached_percent;
  }

  int percent = parse_meminfo();
  if (percent >= 0) {
    cached_percent = percent;
    last_check_ms = now_ms;
  }
  
  return percent;
}

bool is_memory_pressure_low() {
  int percent = get_memory_usage_percent();
  return percent >= 0 && percent < MEMORY_WARNING_THRESHOLD;
}

bool is_memory_pressure_high() {
  int percent = get_memory_usage_percent();
  return percent >= MEMORY_WARNING_THRESHOLD;
}

bool is_memory_pressure_critical() {
  int percent = get_memory_usage_percent();
  return percent >= MEMORY_CRITICAL_THRESHOLD;
}

bool is_memory_pressure_emergency() {
  int percent = get_memory_usage_percent();
  return percent >= MEMORY_EMERGENCY_THRESHOLD;
}

bool should_skip_filesystem_operation() {
  return is_memory_pressure_critical();
}

}  // namespace MemoryPressure
